### PR TITLE
feat(pipeline) Add task status column to Pipeline/Run List

### PR DESCRIPTION
**Pull request by abhinandan13jan**
Add column Task Status to Pipeline and PipelineRun List

### CODE REVIEW

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+export const TaskStatusToolTip: React.FC<TaskStatusToolTipProps> = (props) => {
+  const { tooltip } = props;
+  return (
+    <div className="odc__task__status">
+      {Object.keys(tooltip).map((status) => {
+        return (
+          <div className="row" key={status}>
+            <div className="col-lg-1 col-md-1 offset-lg-1 offset-md-1">
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Use a fixed padding to indent the content instead of using offsets.
Shouldn't you just use `col-*` so that there is never any stacking?

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+interface PipelineTaskStatusProps {
+  pipelinerun: PipelineRun;
+}
+
+interface TaskStatusToolTipProps {
+  tooltip: Tooltip;
+}
+
+interface Tooltip {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: This should be the same type returned from `getTaskStatus` and it should have a name other than `Tooltip` if it's coming from a non-ui utility.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+const getMessage = (status: string, value: number = 0): string => {
+  if (status === 'Notstarted' || status === 'FailedToStart') {
+    return mapColorMessage[status].message;
+  }
+  return `${value} ${mapColorMessage[status].message}`;
+};
+
+export const TaskStatusToolTip: React.FC<TaskStatusToolTipProps> = (props) => {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Move to separate component file.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+  const { tooltip } = props;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Use deconstruction in the function parameters

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import { Tooltip } from '@console/internal/components/utils/tooltip';
+import { getTaskStatus, PipelineRun } from '../../utils/pipeline-augment';
+import '../pipelines/PipelineVisualizationTask.scss';
+
+interface PipelineTaskStatusProps {
+  pipelinerun: PipelineRun;
+}
+
+interface TaskStatusToolTipProps {
+  tooltip: Tooltip;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Should be the same type returned from `getTaskStatus`. Also a more descriptive name than `tooltip` would be useful.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+};
+
+export const PipelineTaskStatus: React.FC<PipelineTaskStatusProps> = (props) => {
+  const { bars, tooltip } = getTaskStatus(props.pipelinerun);
+  return (
+    <div>
+      <Tooltip
+        content={[<TaskStatusToolTip key={Math.floor(100 * Math.random())} tooltip={tooltip} />]}
+      >
+        <VictoryStack horizontal height={35} padding={0} domainPadding={0}>
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Height `35` isn't a standard PF4 size. Probably `32` or `40`.

File: frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationTask.scss [hunk 7]
```diff
@@ -114,3 +114,13 @@ $gutter: 1.8em;
     }
   }
 }
+.odc__task__status {
+  .tooltip__legend {
+    height: 1rem;
+    width: 1rem;
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationTask.scss]: Should use PF4 variables for sizing.

File: frontend/packages/dev-console/src/components/pipelines/PipelineRow.tsx [hunk 8]
```diff
@@ -54,11 +55,13 @@ const PipelineRow: React.FC<PipelineRowProps> = ({ obj, index, key, style }) =>
...
-      <TableData className={tableColumnClasses[3]}>-</TableData>
+      <TableData className={tableColumnClasses[3]}>
+        {(obj.latestRun && <PipelineTaskStatus pipelinerun={obj.latestRun} />) || '-'}
+      </TableData>
       <TableData className={tableColumnClasses[4]}>
-        <Timestamp
-          timestamp={obj.latestRun && obj.latestRun.status && obj.latestRun.status.completionTime}
-        />
+        {obj.latestRun && obj.latestRun.status && obj.latestRun.status.completionTime && (
+          <Timestamp timestamp={obj.latestRun.status.completionTime} />
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelines/PipelineRow.tsx]: It is showing some past date '14 July 1970' or something similar for 'undefined' values

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 9]
```diff
@@ -143,3 +146,38 @@ export const augmentRunsToData = (
...
+      } else {
+        bars.push('Notstarted');
+        tooltip['Notstarted'] ? tooltip['Notstarted']++ : (tooltip['Notstarted'] = 1);
+      }
+    });
+  } else if (
+    pipelinerun &&
+    pipelinerun.status &&
+    pipelinerun.status.conditions &&
+    pipelinerun.status.conditions[0].status === 'False'
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: This is checking for the 'Pipeline failed to Start' criteria which is always in the first condition. i.e before any `taskRun `starts

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+          {bars.map((bar, i) => {
+            return (
+              <VictoryBar
+                key={Math.floor(100 * Math.random())}
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: `bar` is not unique, I tried `bar + i` where react throws error.

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 11]
```diff
@@ -143,3 +146,38 @@ export const augmentRunsToData = (
...
+  ) {
+    bars.push('FailedToStart');
+    tooltip['FailedToStart'] ? tooltip['FailedToStart']++ : (tooltip['FailedToStart'] = 1);
+  } else {
+    bars.push('Notstarted');
+    tooltip['Notstarted'] ? tooltip['Notstarted']++ : (tooltip['Notstarted'] = 1);
+  }
+  return { bars, tooltip };
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: 1 bar = 1 `taskRun` tooltip stacks similar bars. Like bars = ['Succeeded', `Running`, `Succeeded`] will result in tooltip = { Succeded : 2 , Running : 1 }

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+                data={[{ x: 2, y: 1 }]}
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Perhaps I am getting this wrong,

My current implementation is Green: Task1 -> Red:Task2 -> Green: Task3, 
So, all are same width, 1 bar = 1 Task with color representing state and in sequence of startTime of TaskRuns.

Should it be Green: Task1 + Task3 -> Red: Task 2,
In that case what would be the sequence? Green -> Red -> Blue

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx [hunk 13]
```diff
@@ -42,7 +43,10 @@ const PipelineRunRow: React.FC<PipelineRunRowProps> = ({ obj, index, key, style
       <TableData className={tableColumnClasses[2]}>
         <StatusIconAndText status={pipelineRunFilterReducer(obj)} />
       </TableData>
-      <TableData className={tableColumnClasses[3]}>-</TableData>
+      <TableData className={tableColumnClasses[3]}>
+        {' '}
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: `yarn lint` treats some new-lines/spaces as `string` and throws error "All strings must use single quotes". This was included by `yarn lint --fix` :)

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    </div>
+  );
+};
+
+export const PipelineTaskStatus: React.FC<PipelineTaskStatusProps> = (props) => {
+  const { bars, tooltip } = getTaskStatus(props.pipelinerun);
+  return (
+    <div>
+      <Tooltip
+        content={[<TaskStatusToolTip key={Math.floor(100 * Math.random())} tooltip={tooltip} />]}
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: React is throwing error 'All children must have keys' unless I add key to '<TaskStatusToolTip />', perhaps this is because of the [] which holds it. It is the convention to provide Component in [] to tooltip in Console

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 15]
```diff
@@ -143,3 +146,38 @@ export const augmentRunsToData = (
   }
   return data;
 };
+
+export const getTaskStatus = (pipelinerun: PipelineRun) => {
+  let bars = [];
+  let tooltip = {};
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: Dint get that

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: I suspect you actually do not want a space here. So remove it?

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Oh i missed that this is an array. Why is it an array? Doesn't look like there's any valid reason for it to be. 
Don't pass content as an array and you won't get the warning.

File: frontend/packages/dev-console/src/components/pipelines/PipelineRow.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelines/PipelineRow.tsx]: That would happen if `completionTime` is null or 0 instead of undefined.
That's unfortunate.
If you want a dash you'll have to || it.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: I misunderstood how you were generating the chart. I see now how bar won't be unique. 
You should include the task name in your data structure and use it as the key.

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: ok

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: Tooltip is just a bad name. Your utility should be about data structures and not tailoring to a "tooltip".

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: I see now. Hmm. Is there a UX doc describing this behavior?

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx (hunk 13 above)
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineRunRow.tsx]: removed

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 16]
```diff
@@ -0,0 +1,87 @@
...
+interface Tooltip {
+  Succeeded?: number;
+  Running?: number;
+  Failed?: number;
+  Notstarted?: number;
+  FailedToStart?: number;
+}
+
+//The colors are not final. pf-colors are not supported in the charts
+export const mapColorMessage = {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: moved

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: added a static key, It isn't working without the []

[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: @abhinavdahiya The array isn't necessary. Please try again. I even pulled your PR and made the local change successfully.

[the remark this answers is not in the document]
Comment (abhinavdahiya) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: :confused: what PR and what list is this about?

[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: @abhinavdahiya  Sorry, you were mis-commented into this conversation :)

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+        <VictoryStack horizontal height={32} padding={0} domainPadding={0}>
+          {Object.keys(mapColorMessage).map((bar) => {
+            return (
+              taskStatus[bar] > 0 && (
+                <VictoryBar
+                  key={bar}
+                  data={[{ x: 2, y: taskStatus[bar] }]}
+                  barRatio={0.9}
+                  barWidth={32}
+                  padding={{ right: 15 }}
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Where is this being applied? Looks the same without this value.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+    width: 1rem;
+    margin-bottom: 0.5rem;
+  }
+  .tooltip--text {
+    text-align: left;
+  }
+  .tooltip--col {
+    padding-left:  var(--pf-global--spacer--sm);
+  }
+  .bg--running {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss]: You can use the short modifier format although i think you're better off applying the colors directly in code using the same patternfly react-token values used by the charts.
```suggestion
  &.is-running {
```

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 14 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Sorry my mistake!

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+interface PipelineTaskStatusProps {
+  pipelinerun: PipelineRun;
+}
+
+export const PipelineTaskStatus: React.FC<PipelineTaskStatusProps> = ({ pipelinerun }) => {
+  const taskStatus = getTaskStatus(pipelinerun);
+  return (
+    <div>
+      <Tooltip content={[<TaskStatusToolTip key="tooltip" taskStatus={taskStatus} />]}>
+        <VictoryStack horizontal height={32} padding={0} domainPadding={0}>
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: its default responsive behavior of the charts

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 17 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: removing this

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss [hunk 20]
```diff
@@ -0,0 +1,27 @@
+@import '../pipelines/PipelineStatusColors.scss';
+
+.odc__task__status {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss]: updated

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+  .tooltip__legend {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss]: updated

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+    height: 1rem;
+    width: 1rem;
+    margin-bottom: 0.5rem;
+  }
+  .tooltip--text {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss]: updated

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 23]
```diff
@@ -93,6 +118,19 @@ export const getResources = (data: PropPipelineData[]): Resource => {
...
+// The colors are not final. pf-colors are not supported in the charts
+export const mapColorMessage: ColorMessage = {
+  Succeeded: { color: 'green', message: 'Succeded', bgClass: 'bg--success' },
+  Failed: { color: '#C9190B', message: 'Failed', bgClass: 'bg--failed' },
+  Running: { color: '#0066cc', message: 'Running', bgClass: 'bg--running' },
+  Notstarted: { color: '#72767B', message: 'Not Started Yet', bgClass: 'bg--idle' },
+  FailedToStart: {
+    color: '#C9190B',
+    message: 'PipelineRun Failed To Start',
+    bgClass: 'bg--failed',
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: now using pf/react-tokens

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 24]
```diff
@@ -93,6 +118,19 @@ export const getResources = (data: PropPipelineData[]): Resource => {
...
 };
 
+// The colors are not final. pf-colors are not supported in the charts
+export const mapColorMessage: ColorMessage = {
+  Succeeded: { color: 'green', message: 'Succeded', bgClass: 'bg--success' },
+  Failed: { color: '#C9190B', message: 'Failed', bgClass: 'bg--failed' },
+  Running: { color: '#0066cc', message: 'Running', bgClass: 'bg--running' },
+  Notstarted: { color: '#72767B', message: 'Not Started Yet', bgClass: 'bg--idle' },
+  FailedToStart: {
+    color: '#C9190B',
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: using pf tokens

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 25]
```diff
@@ -93,6 +118,19 @@ export const getResources = (data: PropPipelineData[]): Resource => {
   return { propsReferenceForRuns: null, resources: null };
 };
 
+// The colors are not final. pf-colors are not supported in the charts
+export const mapColorMessage: ColorMessage = {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: renamed

File: frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationTask.scss [hunk 26]
```diff
@@ -22,19 +24,19 @@ $gutter: 1.8em;
   }
 
   &.is-done {
-    color: var(--pf-global--success-color--200);
+    color: $pipeline-success;
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelines/PipelineVisualizationTask.scss]: updated `mapRunStatusColor` to be used by both `PipelineRunVisualisation`, `PipelineTaskStatus` and `TaskStatusTooltip` component.

File: frontend/packages/dev-console/src/components/pipelines/PipelineStatusColors.scss [hunk 27]
```diff
@@ -0,0 +1,4 @@
+$pipeline-success: var(--pf-global--success-color--200);
+$pipeline-running: #0066cc;
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelines/PipelineStatusColors.scss]: file deleted

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 28]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  return `${value} ${mapColorMessage[status].message}`;
+};
+
+const TaskStatusToolTip: React.FC<TaskStatusToolTipProps> = ({ taskStatus }) => {
+  return (
+    <div className="odc__task__status">
+      {Object.keys(taskStatus).map((status) => {
+        return (
+          taskStatus[status] && (
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]: using ternary if-else now

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 29]
```diff
@@ -0,0 +1,36 @@
...
+            <div className="row" key={status}>
+              <div className="col-lg-2 col-md-2 tooltip--col">
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]: using a flex layout now

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx (hunk 19 above)
[the remark this answers is not in the document]
Comment (christianvogt) [reply frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.tsx]: Discussed with @serenamarie125 and she would like the height to be a constant 1em and vertically aligned middle with the text so it looks like this:

image

We don't need to block the PR on this particular change though.

File: frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss [hunk 30]
```diff
@@ -0,0 +1,16 @@
+.odc-pipeline-task-status {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/PipelineTaskStatus.scss]: After looking at the layout and how the tooltip contents weren't centered, i think this can be changed to use a grid-layout:

```
.odc-pipeline-task-status {
  display: inline-grid;
  grid-gap: var(--pf-global--spacer--sm);
  text-align: left;
  grid-template-columns: 1rem auto;

  &__legend {
    height: 1rem;
    width: 1rem;
  }
}
```

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 31]
```diff
@@ -0,0 +1,37 @@
...
+    return mapRunStatusColor[status].message;
+  }
+  return `${value} ${mapRunStatusColor[status].message}`;
+};
+
+const TaskStatusToolTip: React.FC<TaskStatusToolTipProps> = ({ taskStatus }) => {
+  return (
+    <div className="odc-pipeline-task-status">
+      {Object.keys(taskStatus).map((status) => {
+        return taskStatus[status] ? (
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]: If you use a grid layout, you should be able to simplify too this:
```
        return taskStatus[status] ? (
          <React.Fragment key={status}>
            <div
              className="odc-pipeline-task-status__legend"
              style={{ background: mapRunStatusColor[status].pftoken.value }}
            />
            <div>{getMessage(status, taskStatus[status])}</div>
          </React.Fragment>
        ) : null;
```

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 32]
```diff
@@ -143,3 +187,31 @@ export const augmentRunsToData = (
   }
   return data;
 };
+
+export const getTaskStatus = (pipelinerun: PipelineRun): TaskStatus => {
+  const taskStatus = { Succeeded: 0, Running: 0, Failed: 0, Notstarted: 0, FailedToStart: 0 };
+  if (pipelinerun && pipelinerun.status && pipelinerun.status.taskRuns) {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/pipeline-augment.ts]: Looks like the status of the pipelinerun doesn't always represent the full set of tasks.
For example in your sample, there are 5 tasks:
image

Yet when this is re-run, it never shows those that have not yet started. Here it says 1 is running, but doesn't show that 4 have not yet started:
image

As the it progresses, it does flash not started for the next task before showing the status of the next set of tasks to run:
image

Finally once complete it shows 5 succeeded.

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 33]
```diff
@@ -0,0 +1,37 @@
+import * as React from 'react';
+import { TaskStatus, mapRunStatusColor } from '../../utils/pipeline-augment';
+
+interface TaskStatusToolTipProps {
+  taskStatus: TaskStatus;
+}
+
+const getMessage = (status: string, value: number = 0): string => {
+  if (status === 'Notstarted' || status === 'FailedToStart') {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]: We should look at creating an enum for the known status values. Not sure if existing pipelines code has these strings. If so we can do a separate PR to track the refactor.

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts (hunk 32 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: The problem with this is that the `PipelineRun` data doesn't have any *details/name* of `Pending/Future` tasks. Only tasks in Failed, Succeeded, `Running` and `Starting/About to Start` `conditions` are present in the `PipelineRun`. For listing `Pending` tasks, the reference `Pipeline` must be examined using `TaskRef` and any `Tasks` which are present in the reference `Pipeline` but not the `PipelineRun` could be marked as `Pending`. This should be a separate PR/Effort as this would require to bring in the corresponding `Pipeline` *(Firehose)* details in the `PipelineRun` List.

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx (hunk 33 above)
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]: No, we don't have strings for these values, it's better to have a common enum

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts (hunk 32 above)
Comment (christianvogt) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: But this is also a general problem with tekton and the pipeline run not having enough data in it.
I'm fine if we take this up as a bug.

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 34]
```diff
@@ -0,0 +1,35 @@
+import * as React from 'react';
+import { TaskStatus, runStatus, getRunStatusColor } from '../../utils/pipeline-augment';
+import './TaskStatusToolTip.scss';
+
+interface TaskStatusToolTipProps {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]:
```suggestion
interface TaskStatusTooltipProps {
```

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+const getMessage = (message: string, value: number = 0): string => {
+  if (message === runStatus.PipelineNotStarted || message === runStatus.FailedToStart) {
+    return message;
+  }
+  return `${value} ${message}`;
+};
+
+const TaskStatusToolTip: React.FC<TaskStatusToolTipProps> = ({ taskStatus }) => {
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]:
```suggestion
const TaskStatusTooltip: React.FC<TaskStatusTooltipProps> = ({ taskStatus }) => {
```

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+  return (
+    <div className="odc-pipeline-task-status">
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]:
```suggestion
    <div className="odc-task-status-tooltip">
```

File: frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx [hunk 37]
```diff
@@ -0,0 +1,35 @@
...
+      {Object.keys(runStatus).map((status) => {
+        let {message,pftoken} = getRunStatusColor(status);
+        return taskStatus[status] ? (
+          <React.Fragment key={status}>
+            <div
+              className="odc-pipeline-task-status__legend"
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/components/pipelineruns/TaskStatusToolTip.tsx]:
```suggestion
              className="odc-task-status-tooltip__legend"
```

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 38]
```diff
@@ -1,15 +1,38 @@
+/* eslint-disable @typescript-eslint/camelcase */
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/pipeline-augment.ts]: Do not disable eslint rules. 
You can use
```
import { chart_color_green_400 as successColor } from ...
```

File: frontend/packages/dev-console/src/utils/pipeline-augment.ts [hunk 39]
```diff
@@ -143,3 +168,73 @@ export const augmentRunsToData = (
...
+      return { message: 'Cancelled', pftoken: chart_color_black_400 };
+    default:
+      return { message: 'PipelineRun Not started Yet', pftoken: chart_color_black_200 };
+  }
+};
+
+export const getTaskStatus = (pipelinerun: PipelineRun): TaskStatus => {
+  let taskStatus;
+  Object.keys(runStatus).forEach((key) => {
+    taskStatus = { ...taskStatus, key };
```
Comment (christianvogt) [inline frontend/packages/dev-console/src/utils/pipeline-augment.ts]:
```
  const taskStatus: TaskStatus = {};
  Object.keys(runStatus).forEach((key) => {
    taskStatus[key] = 0;
  });
```

Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/utils/pipeline-augment.ts]: this is throwing error, cannot set value `Success` (first key) of undefined

**Comment by spadgett** [comment]
/hold

We're having infrastructure problems from the CI outage earlier today. Holding for now so the test namespace can be cleaned up. (I believe this takes an hour.) Feel free to remove the hold later.

> error: could not run steps: step e2e-aws failed: could not wait for template instance to be ready: could not determine if template instance was ready: failed to create objects: object is being deleted: pods "e2e-aws" already exists
